x/ibc-hooks/contracts/crosschain-swaps/src/execute.rs

```diff
@@ -0,0 +1,271 @@
+        let channel_id = swap_msg_state.forward_to.channel;
+        let to_address = swap_msg_state.forward_to.receiver;
+        let data = CrosschainSwapResponse {
+            msg: format!("Sent {amount}{denom} to {channel_id}/{to_address}"),
+        };
+
+        return Ok(response
+            .set_data(to_binary(&data)?)
+            .add_message(ibc_transfer));
+    }
+
+    // Check that there isn't anything stored in FORWARD_REPLY_STATES. If there
+    // is, it means that the contract is already waiting for a reply and should
+    // not override the stored state. This should never happen here, but adding
+    // the check for safety. If this happens there is likely a malicious attempt
+    // modify the contract's state before it has replied.
+    if FORWARD_REPLY_STATES.may_load(deps.storage)?.is_some() {
+        return Err(ContractError::ContractLocked {
+            msg: "Already waiting for a reply".to_string(),
+        });
+    }
+    // Store the ibc send information and the user's failed delivery preference
+    // so that it can be handled by the response
+    FORWARD_REPLY_STATES.save(
+        deps.storage,
+        &ForwardMsgReplyState {
+            channel_id: swap_msg_state.forward_to.channel,
+            to_address: swap_msg_state.forward_to.receiver.into(),
+            amount: swap_response.amount.into(),
+            denom: swap_response.token_out_denom,
+            failed_delivery: swap_msg_state.forward_to.failed_delivery,
+        },
+    )?;
+
+    Ok(response.add_submessage(SubMsg::reply_on_success(ibc_transfer, FORWARD_REPLY_ID)))
+}
+
+use ::prost::Message; // Proveides ::decode() for MsgTransferResponse
+
+pub fn handle_forward_reply(deps: DepsMut, msg: Reply) -> Result<Response, ContractError> {
+    // Parse the result from the underlying chain call (IBC send)
+    let SubMsgResult::Ok(SubMsgResponse { data: Some(b), .. }) = msg.result else {
+        return Err(ContractError::FailedIBCTransfer { msg: format!("failed reply: {:?}", msg.result) })
+    };
+
+    // The response contains the packet sequence. This is needed to be able to
+    // ensure that, if there is a delivery failure, the packet that failed is
+    // the same one that we stored recovery information for
+    let response =
+        MsgTransferResponse::decode(&b[..]).map_err(|_e| ContractError::FailedIBCTransfer {
+            msg: format!("could not decode response: {b}"),
+        })?;
+
+    // Similar consideration as the warning above. Is it safe for this to be an Item?
+    let ForwardMsgReplyState {
+        channel_id,
+        to_address,
+        amount,
+        denom,
+        failed_delivery,
+    } = FORWARD_REPLY_STATES.load(deps.storage)?;
+    FORWARD_REPLY_STATES.remove(deps.storage);
+
+    // If a recovery address was provided, store sent IBC transfer so that it
+    // can later be recovered by that addr.
+    if let Some(Recovery { recovery_addr }) = failed_delivery {
```

Comment:
I don't understand what is this `failed_delivery` argument were parsing? Seems like were marking this as in-flight, not failed?

Reply:
failed_delivery comes from:

```
#[cw_serde]
pub struct ForwardTo {
    pub channel: String,
    pub receiver: Addr,
    pub next_memo: Option<String>,
    pub failed_delivery: Option<Recovery>,
}

#[cw_serde]
pub struct SwapMsgReplyState {
    pub swap_msg: SwapRouterExecute,
    pub contract_addr: Addr,
    pub block_time: Timestamp,
    pub forward_to: ForwardTo,
}
```

It represents what to do on a failed delivery (right now that's a recovery addr, but I can imagine it being more complex). I can see how it is a confusing name though. Maybe something like:

```
enum FailedIBCForwardAction {
   None,
   LocalRecovery { addr: Addr }
   ...
}
```
(and remembering to rename IBC so it doesn't serialize to `i_b_c` 😓 )